OSTEP H2

tbonelee opened the pull request:
OSTEP 실습 [H2](https://github.com/remzi-arpacidusseau/ostep-projects/tree/master/scheduling-xv6-lottery) 과제입니다.

# 구현 사항

- 시스템콜 `int getpinfo(struct pstat*);` 추가
- 시스템콜 `int settickets(int);` 추가
- 추첨 스케쥴링 구현
- 스케쥴링 테스트용 커맨드 `h2graph` 추가
  -  30:20:10의 비율로 티켓 설정하여 시뮬레이션 돌리고 그래프 출력
    <img width="1341" height="617" alt="image" src="https://github.com/user-attachments/assets/96f7a86b-f110-4650-902a-f3688459e0c3" />

# 이슈?
1. 모든 자식 프로세스가 `sleep()` 호출 상태일 때 ticks 비율이 1:1:1로 스케쥴링되는 현상 (`settickets()` 처리되게 하기 위해 자식 프로세스 초기에 `sleep()` 호출 길게 줬을 때 발생)
  - 파악해본 내용 :
    - `clockintr()` -> `wakeup()`에서 잠들어있던 모든 프로세스를 잠시 깨움.
    - 깨어난 프로세스는 모두 RUNNABLE 상태가 되고 이 상태에서 스케쥴러(`scheduler()`)가 호출되어 하나씩 스케쥴링하고, 각 프로세스의 `ticks`가 증가
    - 스케쥴러가 (`sleep()`호출 상태인) RUNNABLE 프로세스를 스케쥴링해서 실행하면 `sys_sleep()`의 while 루프 초기로 돌아가서 다시 잠듦.
    - 모든 자식 프로세스가 한 번씩 실행되고 잠든 다음 다시 다음 타이머 인터럽트까지 대기.
2. CPU가 자식 프로세스 갯수 이상으로 존재하면 프로세서 자원 획득 위해 경쟁할 필요가 없으므로 1:1:1의 비율로 스케쥴링되는 현상

tbonelee committed:
`proc` 구조체에 티켓 속성 추가

tbonelee committed:
PRNG 추가

tbonelee committed:
pstat.h 추가

tbonelee committed:
`int settickets(int)` 시스템콜 추가

tbonelee committed:
`int getpinfo(struct pstat*)` 추가

tbonelee committed:
스케쥴링 방식 Lottery 스케쥴링으로 변경

tbonelee committed:
proc 구조체 tickets, ticks 락 획득 누락 점검

tbonelee committed:
proc.tickets, proc.ticks 자료형 변경

tbonelee committed:
proc 접근 이후에 락 획득하는 버그 수정

tbonelee committed:
시뮬레이션으로 누적 ticks 기록하는 쉘 커맨드 `h2graph` 추가

tbonelee committed:
불필요한 공백 변경 이전 상태로 리포맷

simnalamburt commented on kernel/proc.c:
```diff
@@ -457,27 +469,56 @@ scheduler(void)
 
-    int found = 0;
-    for(p = proc; p < &proc[NPROC]; p++) {
+    uint tickets_for_runnable[NPROC];
```
이친구가 쫌 거대해서 약간 마음에 걸리네용 이친구 없이 짜면 어떨까요?

simnalamburt commented on kernel/proc.c:
```diff
@@ -457,27 +469,56 @@ scheduler(void)
-    for(p = proc; p < &proc[NPROC]; p++) {
+    uint tickets_for_runnable[NPROC];
+
+    int total_tickets = 0;
```
uint

tbonelee replied:
티켓 관련 값은 uint32로 통일했습니다

simnalamburt commented on kernel/proc.c:
```diff
@@ -457,27 +469,56 @@ scheduler(void)
+    uint tickets_for_runnable[NPROC];
+
+    int total_tickets = 0;
+    for(int i = 0; i < NPROC; i++) {
```
uint

tbonelee replied:
uint로 수정했습니다

simnalamburt commented on kernel/pstat.h:
```diff
@@ -0,0 +1,13 @@
+  int inuse[NPROC];   // whether this slot of the process table is in use (1 or 0)
+  int tickets[NPROC]; // the number of tickets this process has
+  int pid[NPROC];     // the PID of each process 
+  int ticks[NPROC];   // the number of ticks each process has accumulated 
```
uint64

tbonelee replied:
ticks 관련 값은 uint64로 통일했습니다

simnalamburt commented on kernel/pstat.h:
```diff
@@ -0,0 +1,13 @@
+
+struct pstat {
+  int inuse[NPROC];   // whether this slot of the process table is in use (1 or 0)
+  int tickets[NPROC]; // the number of tickets this process has
```
uint

tbonelee replied:
수정했습니다

simnalamburt commented on kernel/rand.c:
```diff
@@ -0,0 +1,21 @@
+// start is inclusive, end is exclusive
+int
+getrandom(int start, int end) {
+  return start + rand() % (end - start);
```
Not uniform
https://github.com/simnalamburt/snippets/blob/main/python/uniformness-of-random/main.ipynb

tbonelee replied:
오.. 생각지 못했던 부분이네요

simnalamburt commented on kernel/rand.h:
```diff
@@ -0,0 +1 @@
+int getrandom(int start, int end);
```
Use defs.h instead or make it static

tbonelee replied:
defs.h로 옮겼습니다

simnalamburt commented on kernel/syscall.c:
```diff
@@ -102,6 +102,8 @@ extern uint64 sys_link(void);
 extern uint64 sys_close(void);
 extern uint64 sys_getreadcount(void);
+extern uint64 sys_settickets(void);
+extern uint64 sys_getpinfo(void);
```
Type

simnalamburt commented on kernel/sysproc.c:
```diff
@@ -91,3 +92,32 @@ sys_uptime(void)
   return xticks;
 }
+
+uint64
```
Why uint64?

tbonelee replied:
다른 시스템 콜도 음수를 반환하지만 `uint64`로 썼길래 컨벤션 같은 거로 생각하고 `uint64`로 했었습니다!
`int`가 더 자연스러운가요 혹시??

simnalamburt replied:
unsigned인게 문제가 되기보다는, 큰 값을 반환하지 않고 무조건 0 아니면 -1 반환하는 시스콜이라 굳이 큰 타입을 쓸 필요는 없는거같아요

tbonelee replied:
아하 넵 수정했습니다!

tbonelee replied:
시스템 콜 함수 전부 담고 있는 배열이 `uint64 (*)(void)`의 배열이라 컴파일 오류가 나서 어쩔 수 없이 다시 복구 시켰습니다🥲

simnalamburt replied:
아 그러네요

simnalamburt reviewed (changes requested):
스트라이드 스케줄러도 한번 구현해보시면 좋을것같아용

tbonelee committed:
랜덤 함수 수정

- modulo bias 완화
- 알고리즘 변경 : 기본 LCG -> xoshiro128++
- 초기 시드값 하드웨어에서 가져오기
- splitmix64 알고리즘 사용하여 랜덤 시드 배열 초기화

tbonelee committed:
unsigned 가능한 곳은 타입 수정

tbonelee committed:
rand.h를 defs.h로 통합

tbonelee committed:
공평하지 않은 선택 조건 수정

tbonelee committed:
`int getpinfo()` 선언된 헤더 수정(proc.h -> defs.h)

tbonelee committed:
sys_settickets, sys_getpinfo 리턴 타입 수정

tbonelee committed:
Stride 스케쥴러로 변경

tbonelee committed:
Update log message in h2graph.c for clarity

tbonelee committed:
Update return types for sys_settickets and sys_getpinfo to int

tbonelee commented:
@simnalamburt 리뷰해주신 내용 반영하고 스트라이드로 수정해보았습니다

tbonelee committed:
Fix incompatible pointer type

simnalamburt commented on user/h2graph.c:
```diff
@@ -0,0 +1,149 @@
+             * 3. 스케쥴러가 하나씩 RUNNABLE한 프로세스를 실행하면, 그때마다 하나씩 다시 잠들어서 모든 프로세스가 한 번씩 ticks가 증가
+             * 4. 모두 잠들고 나서야 다시 타이머 인터럽트 발생.
+             */
+            /* 이슈2) CPUS=3 이상부터는 프로세서 자원을 사용하기 위해 경쟁할 필요가 없으므로 ticks 비율이 1:1:1이 되는 현상?
```
이건 이슈는 아니고 정상적인 구현인거같습니다

simnalamburt commented on user/h2graph.c:
```diff
@@ -0,0 +1,149 @@
+             */
+            /* 이슈2) CPUS=3 이상부터는 프로세서 자원을 사용하기 위해 경쟁할 필요가 없으므로 ticks 비율이 1:1:1이 되는 현상?
+            */
+            printf("%d\t", measurement * SLEEP_INTERVAL);
+            for (int i = 0; i < NUM_PROCESSES; i++) {
+                printf("%d\t\t", ticks[i]);
+            }
+            for (int i = 0; i < NUM_PROCESSES; i++) {
+                printf("%d\t\t", found_tickets[i]);
+            }
+            for (int i = 0; i < NUM_PROCESSES; i++) {
+                print_ratio2(ratios[i]);
+                if (i < NUM_PROCESSES - 1) printf("\t");
+            }
+            printf("\n");
```
그래프 예쁘네요 근데 다음부터는 C 프로그램으로는 숫자만 찍고 그래프는 엑셀이나 파이썬같은거 써서 외부프로그램으로 그리셔도 됩니다

simnalamburt commented on kernel/rand.c:
```diff
@@ -0,0 +1,71 @@
+void
+randinit(void) {
+  seed_rand(r_time());
+}
```
스트라이드 스케줄링으로 바꾸면서 호출하는 장소가 없어진것같네요. 없애거나 호출하는곳을 만드는게 좋을것같습니다.

tbonelee replied:
일단 제거하고 필요할 때 복구하는 쪽으로 가보겠습니다

simnalamburt commented on kernel/rand.c:
```diff
@@ -0,0 +1,71 @@
+
+	s[2] ^= t;
+
+	s[3] = rotl(s[3], 11);
```
랜덤 state `uint32 s[4]` 는 전역변수여서, s를 수정하는 코드를 작성하면 s는 shared mutable state가 되고, 여러 CPU가 `get_random_below()` 를 동시에 호출하면 데이터레이스가 발생합니다.

(뒤집어 말하면 s가 전역변수여도 s를 읽기만 한다면 별도로 s를 보호하지 않아도 된다는 뜻이기도 합니다)

데이터레이스에 대한 몇가지 해결책이 있는데

1. 락으로 보호하거나 (가장 안좋은 해결책)
2. `uint32 s[4]` 를 `_Atomic uint128 s` 로 만들고 s에 write할때에도 딱 한번만 write 하도록 수정하기
3. s `struct cpu` 안으로 옮겨서, CPU마다 각자 자신만의 랜덤 state를 갖게합니다. 어차피 하나의 rand state를 공유하나 각자 rand state를 따로 가지나, 여러 CPU의 rand state가 같은값에서 출발하지만 않는다면 CPU별로 출력하는 랜덤값은 모두 유니폼할것이기 때문입니다. 이러면 rand state가 비록 mutable하더라도 shared가 아니므로, shared mutable state가 아니어서 보호할 필요가 없어집니다

저는 3번이 가장 나은 해결책같아보입니다. 지금의 케이스에선 이 코드를 아예 없애는게 가장 나아서 이 테크닉을 쓸 일이 없어보이지만, 나중에 쓰셔요.

tbonelee replied:
여러 방안들에 대한 아이디어 감사합니다. 레이스 컨디션에 대해 한 번씩 더 체크해보는 습관을 들여야겠네요..!

simnalamburt commented on kernel/pstat.h:
```diff
@@ -0,0 +1,14 @@
+#include "param.h"
+#include "types.h"
+
+struct pstat {
+  int inuse[NPROC];   // whether this slot of the process table is in use (1 or 0)
+  uint32 tickets[NPROC]; // the number of tickets this process has
+  int pid[NPROC];     // the PID of each process 
+  uint64 ticks[NPROC];   // the number of ticks each process has accumulated 
+};
```
`int inuse[NPROC];` 대신 `_Bool inuse[NPROC];` 을 쓰면 더 작아질것같습니다.

tbonelee replied:
아하 저런 식으로 쓸 수 있군요..! 수정하겠습니다

simnalamburt commented on kernel/proc.h:
```diff
@@ -104,4 +108,8 @@ struct proc {
   struct file *ofile[NOFILE];  // Open files
   struct inode *cwd;           // Current directory
   char name[16];               // Process name (debugging)
+
+
```
사소한것이긴 한데 PR submit 전에 이런 불필요한 whitespace 수정은 미리 없애주면 가오가 삽니다

tbonelee replied:
명심하겠습니다 👍

simnalamburt commented on kernel/proc.h:
```diff
@@ -104,4 +108,8 @@ struct proc {
+
 };
+
+#define MAX_TICKETS 10000
```
proc.c 에서밖에 안쓰는 매크로여서, 저라면 proc.c 안에 정의했을것같습니다

tbonelee replied:
일리 있다고 생각해서 수정하겠습니다!

simnalamburt commented on kernel/proc.c:
```diff
@@ -169,6 +170,36 @@ freeproc(struct proc *p)
+// 0으로 초기화되어도 `get_runnable_min_pass_proc`에서 적절한 값으로 덮어씌워진다.
+// 프로그램 초기에 덮어씌워지지 않은 값이 읽혀서 사용되더라도 프로그램 초기에는
+// 실제 최소 pass 값이 0이므로 큰 이슈가 없을 것으로 예상?
+static uint64 cached_min_pass = 0;
```
shared mutable state여서 락이든 아토믹이든 무언가에 의해 보호되어야합니다

tbonelee replied:
_Atomic으로 수정해보겠습니다

simnalamburt commented on kernel/proc.c:
```diff
@@ -457,28 +504,38 @@ scheduler(void)
-    }
-    if(found == 0) {
+
+    // 고민) 멀티프로세서 환경에서 다른 스케쥴러가 proc 락을 acquire한 경우
+    //      해당 프로세서 타이머 인터럽트 때까지 get_runnable_min_pass_proc()은 블로킹??
+    p = get_runnable_min_pass_proc();
+
+    if(p == 0) {
       // nothing to run; stop running on this core until an interrupt.
       asm volatile("wfi");
+      continue;
     }
+
+    // min_pass_proc을 뽑고난 후 다시 실행시키는 사이에 다른 프로세서가
+    // 상태를 변경할 수 있으므로 락을 잡고 상태를 확인?
+    acquire(&p->lock);
+    if (p->state == RUNNABLE) {
```
합당한, 좋은 고민을 하셨고, 해결방법이 있습니다.

`get_runnable_min_pass_proc()` 의 구현을 보면, 최소값을 찾는 과정에서 p의 락을 잡았다가 풀게되어있습니다. 그리고 `get_runnable_min_pass_proc()` 함수가 리턴된 뒤 즉시 다시 락을 잡게되어있습니다.

이 문제는, `get_runnable_min_pass_proc()` 함수의 이름을 `get_runnable_min_pass_proc_locked()` 로 바꾸고, 저 함수의 동작도 실행 가능한 최소 pass 값의 proc을 그냥 리턴하는게 아니라 **락이 잡힌 상태에서** 리턴하도록 하면 깔끔하게 해결됩니다.

tbonelee replied:
이런 좋은 방법이 있었군요..! 감사합니다 ㅎㅎㅎ

simnalamburt commented on kernel/proc.h:
```diff
@@ -91,6 +91,10 @@ struct proc {
   int xstate;                  // Exit status to be returned to parent's wait
   int pid;                     // Process ID
+  uint32 tickets;              // Number of tickets for scheduling
+  uint32 stride;               // Stride for Stride Scheduling
```
지금 구현을 보면 프로세스가 유효한 상태라면 `stride == MAX_TICKETS / tickets` 이 invariant가 언제나 항상 성립하는것같습니다. 이 변수는 없앨수있을 것 같습니다.

tbonelee replied:
처음 생각으로는 매번 stride값이 필요할 때마다 `MAX_TICKETS / tickets`연산을 하는 것을 줄여보겠다는 생각으로 추가했었습니다
메모리를 더 차지하는 것에 비해 미미한 효과일까요??
측정을 안 해보고 섣부르게 필드를 추가한 것 같다는 생각도 들긴 하는데.. 혹시 이 정도의 트레이드오프를 판단하시는 경험칙 같은 게 있을까 해서 여쭤봅니다!

tbonelee replied:
일단은 제거해서 올렸습니다!

simnalamburt replied:
제생각에는 변수가 없는쪽이 더 나은거같아요

tbonelee replied:
넵넵 없애는 쪽으로 가보겠습니다

simnalamburt reviewed (changes requested):
지금까지 스터디에서 제출된 H2중 가장 잘짜셨습니다

simnalamburt commented:
참고로 제가 이 프로젝트에 write 권한이 없어서 해결된 코멘트들에 대해 resolve 버튼을 누를 권한이 없습니다

tbonelee committed:
사용되지 않는 랜덤 함수 제거

tbonelee committed:
pstat->inuse의 타입을 _Bool로 변경

tbonelee committed:
proc.c에서만 사용되는 매크로를 proc.c로 이동

tbonelee committed:
_Atomic으로 cached_min_pass 보호

tbonelee committed:
proc->stride 속성 제거

tbonelee committed:
get_runnable_min_pass_proc()이 락을 잡은 채로 반환하도록 수정

simnalamburt commented on kernel/proc.c:
```diff
@@ -320,6 +364,15 @@ fork(void)
+  // For Stride Scheduling
+  // The child process inherits the number of tickets from the parent
+  np->tickets = p->tickets;
+  // 자식 프로세스의 pass값은 가장 작은 pass 값으로 설정.
+  // 자식 프로세스가 부모 프로세스와 같은 pass 값을 갖는 경우 스케쥴링에서 부모와 동일한 선에서 경쟁하게 된다.
+  // 대신 새 자식 프로세스의 pass 값을 글로벌 min pass 값으로 설정하게 되면 최대 한 stride만큼 부모보다 유리한 선에서 스케쥴링
+  // (부모 프로세스 pass - 글로벌 min pass)값이 부모의 stride보다 크다면 부모는 pass값이 최소가 아닐 때 스케쥴링된 것이므로 현재 정책과 모순
+  np->pass = cached_min_pass;
```
리눅스 CFS 와 동일하게 잘 디자인하셨습니다

https://github.com/torvalds/linux/blob/c30a13538d9f8b2a60b2f6b26abe046dea10aa12/Documentation/scheduler/sched-design-CFS.rst#3--the-rbtree

tbonelee replied:
오 저기도 비슷한 식이군요..

simnalamburt replied:
지금 한거에서 min_pass_value/min_pass_proc 검색을 O(n) 에서 O(log n) 내지는 O(1) 로 빠르게 만들고 사소한거 몇개 더 고치면 그게 CFS라고 생각하시면 돼요

simnalamburt commented on kernel/proc.c:
```diff
@@ -169,6 +172,42 @@ freeproc(struct proc *p)
+get_runnable_min_pass_proc_locked(void) {
+  uint64 min_pass = ((uint64)~0ULL); // UINT64_MAX
+  struct proc* min_pass_proc = 0;
+  for(int i = 0; i < NPROC; i++) {
+    acquire(&proc[i].lock);
+    if(proc[i].state == RUNNABLE) {
+      if(proc[i].pass < min_pass) {
+        // 이전 min_pass_proc이 존재한다면 락을 해제  
+        if(min_pass_proc != 0) {
+          release(&min_pass_proc->lock);
+        }
+        min_pass = proc[i].pass;
+        min_pass_proc = &proc[i];
+        // 새 min_pass_proc에 대한 락은 유지
+      } else {
+        release(&proc[i].lock);
+      }
+    } else {
+      release(&proc[i].lock);
+    }
+  }
```
요렇게 짜면 로직이 좀 더 단순합니다

```suggestion
  for(int i = 0; i < NPROC; i++) {
    acquire(&proc[i].lock);

    if(proc[i].state == RUNNABLE && proc[i].pass < min_pass) {
      // 이전 min_pass_proc이 존재한다면 락을 해제
      if(min_pass_proc != 0) { release(&min_pass_proc->lock); }

      min_pass = proc[i].pass;
      min_pass_proc = &proc[i];

      // 새 min_pass_proc에 대한 락을 유지한채로, 다음 프로세스 검사
      continue;
    }

    release(&proc[i].lock);
  }
```

simnalamburt reviewed (approved):
고생많으셨습니다 잘짜셨네요

simnalamburt commented on kernel/proc.c:
```diff
@@ -457,28 +510,31 @@ scheduler(void)
+      continue;
     }
+
+    // Switch to chosen process.  It is the process's job
+    // to release its lock and then reacquire it
+    // before jumping back to us.
+    p->state = RUNNING;
+    c->proc = p;
+    
+    swtch(&c->context, &p->context);
+
+    p->pass += MAX_TICKETS / p->tickets;
+    cached_min_pass = p->pass;
+    p->ticks++;
```
현재 구현에선 같은 min_pass 값을 가진 프로세스가 여러개 있을 수 있어서, 경우에 따라 특정 프로세스가 cached_min_pass보다 낮은 값의 min_pass를 가질수도 있게됩니다. 아래와 같이 min_pass 를 찾은 즉시 cached_min_pass 를 업데이트하도록 하면, cached_min_pass보다 낮은 pass 값을 가지는 프로세스가 없음이 보장됩니다.

```suggestion
    cached_min_pass = p->pass;

    // Switch to chosen process.  It is the process's job
    // to release its lock and then reacquire it
    // before jumping back to us.
    p->state = RUNNING;
    c->proc = p;
    
    swtch(&c->context, &p->context);

    p->pass += MAX_TICKETS / p->tickets;
    p->ticks++;
```

다만 이렇게하면 cached_min_pass가 실제 min_pass와 같지 않고 더 낮은 경우가 있을 수 있습니다. 지금 프로세스 목록이 배열로 관리되고있고, pass로 정렬해놓은 힙이나 트리가 어디에 따로 없어서 min_pass 계산이 O(n)으로 비싸다보니 이런 문제가 생기는건데,

1. cached_min_pass가 순간적으로 실제 min_pass보다 높아지는것을 허용하기 (기존 구현)
2. cached_min_pass가 순간적으로 실제 min_pass보다 낮아지는것을 허용하기 (제 제안)

둘중 하나를 골라야하는데 저는 1번보다는 2번이 나은것같습니다.

tbonelee replied:
아 생각못했던 부분이네요. 저도 새 프로세스 pass 값이 최소값으로 보장되는 게 더 중요하다 생각해서 제안해주신대로 수정하겠습니다!

tbonelee committed:
get_runnable_min_pass_proc_locked() 리팩토링

Co-authored-by: Jihyeon Kim (김지현) <simnalamburt@gmail.com>

tbonelee committed:
cached_min_pass 갱신 시점 변경

- min pass 조회 당시의 pass값으로 갱신하도록 수정

Co-authored-by: Jihyeon Kim (김지현) <simnalamburt@gmail.com>
